new-cep.md

```diff
@@ -0,0 +1,226 @@
+of all network requests in conda. The way we intend to do this is by
+giving plugin authors the ability to replace the [`CondaSession`][conda-session]
+class. To customize it, plugin authors can either subclass our existing session 
+class or write a new class that conforms to the [`requests.Session`][requests-session] 
+API. Furthermore, this can either configured on a global level or on a 
+channel-by-channel basis.
+
+Our primary motivation for creating this new plugin hook is giving conda the
+ability to better support various authentication schemes (e.g. OAuth or HTTP Basic
+Authentication). Included in this CEP is a link to a fully functional prototype
+illustrating how we intend to enable better HTTP Basic Authentication support
+in conda and to illustrate exactly how this new plugin hook is supposed to work.
+We believe that this hook not only serves for better support of various authentication
+schemes right now but is flexible enough to handle other use cases in the future.
+
+## Specification
+
+The "conda fetch" plugin hook gives plugin authors the ability to completely replace the
+[`CondaSession`][conda-session] class currently defined in conda to handle all network 
+traffic. This can either be done on a per channel basis or for all
+network requests. Plugin authors will have the ability to define
+a completely new class as long as it conforms to the 
+[`requests.Session`][requests-session]
+API or simply subclass the existing [`CondaSession`][conda-session] class to make any
+desired customizations.
+
+In order to make it clearer to plugin authors wishing to subclass
+the current [`CondaSession`][conda-session] class, we will be renaming
+it to `CondaRequestsSession` to emphasize that this belongs to the
+requests library. The old name will be deprecated and follow our
+[standard deprecation schedule][conda-deprecation].
+
+Below is an example of how a plugin author would use this new hook
+to create their own custom [`CondaSession`][conda-session] class:
+
+```python
+from conda.gateways.connection.session import CondaRequestsSession
+from conda.plugins import hookimpl, CondaFetch
+
+
+PLUGIN_NAME = "custom_fetch"
+
+
+class CustomSession(CondaRequestsSession):
+    """
+    Our custom CondaRequestsSession class which defines an additional attribute
+    """
+    def __init__(self):
+        super().__init__()
+        self.custom_attr = "custom_attr"
+
+
+@hookimpl
+def conda_session_classes(): 
+    """
+    Register our custom CondaFetch class
+    """
+    yield CondaFetch(
+        name=PLUGIN_NAME, 
+        session=CustomSession
+    )
+```
+
+### Using the plugin
+
+#### Specifying via configuration files
+
+In order to configure this new plugin, users will either configure it
+globally for conda (i.e. all network requests go through this class):
+
+```yaml
+fetch: custom_fetch
+```
+
+or on a per channel basis:
+
+```yaml
+channels:
+    - https://my-custom-conda-packages.com:
+        fetch: custom_fetch
+```
```

Comment:
How can this be achieved via the CLI or env vars?

Reply:
@jaimergp,

I am not sure there is a clean way to do this that would be the most flexible.

Here's an example of a method one could use on the command line to effectively achieve the the configuration defined above:

```bash
conda install python \
  --channels "https://my-custom-conda-packages.com" \
  --fetch custom_fetch \
  --override-channels
```

and with environment variables:

```bash
export CONDA_CHANNELS="https://my-custom-conda-packages.com"
export CONDA_OVERRIDE_CHANNELS=1
export CONDA_FETCH=custom_fetch
conda install new_package
```

The only problem is that this approach breaks as soon as you want to add another channel in there (e.g. "conda-forge"). For the most flexible option, I think passing in JSON as the value for channel makes the most sense:

```bash
conda install python \
  --channels '{"https://my-custom-conda-packages.com": {"fetch": "custom_fetch"}}'
```

and as an environment variable:

```bash
export CONDA_CHANNELS='{"https://my-custom-conda-packages.com": {"fetch": "custom_fetch"}}'
```

That is definitely not the prettiest approach, but it Works™️.

Do you have any other ideas?

Reply:
Yea, after some thinking I think we would need JSON support for maps. I don't know if that can become a problem by being too flexible, though. Nothing Pydantic validation and schemas cannot help with :D